### PR TITLE
fix namespace filter to not jump after clicking

### LINKED ISSUE #851: Namespace multi-selector hover highlighting jumps around
**Describe the bug**

Screen Recording 2020-09-10 at 12 57 19 PM

This was noticed on the `Workloads/Pods` page

**Expected behavior**
After clicking a namespace highlighted namespace doesn't jump. If the mouse is not over any item then none should be highlighted.

**Environment (please complete the following information):**
- Lens Version: 3.6.0-rc2
- OS: [e.g. OSX] macos
- Installation method (e.g. snap or AppImage in Linux): download

### PR DESCRIPTION
**Pull request by Nokel81**
Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes #851

### CODE REVIEW

File: src/renderer/components/+namespaces/namespace-select.tsx [hunk 1]
```diff
@@ -70,38 +71,61 @@ export class NamespaceSelect extends React.Component<Props> {
...
+
   render() {
-    const { contextNs, hasContext, toggleContext } = namespaceStore;
-    let placeholder = <Trans>All namespaces</Trans>;
-    if (contextNs.length == 1) placeholder = <Trans>Namespace: {contextNs[0]}</Trans>
-    if (contextNs.length >= 2) placeholder = <Trans>Namespaces: {contextNs.join(", ")}</Trans>
     return (
-      <NamespaceSelect
-        placeholder={placeholder}
+      <ReactSelect
```
Comment (nevalla) [inline src/renderer/components/+namespaces/namespace-select.tsx:111]: Do we use `NamespaceSelect` component somewhere after this change?

**Review by nevalla — COMMENTED** [review]
Noticed that this PR also in addition to bug fix, changes the behaviour a bit too. First, there's no more "namespace" icons anymore rendered. Also, when selecting a namespace, the background of selected item is changing. Those changes are a bit out of scope of the actual fix. I'm not saying those changes are bad, but I have a bit mixed feelings changing those in this PR.

Since this is touching quite many files and fundamentally changes some parts of the system, I would like to have extra eyes too to check this out.

**Comment by Nokel81** [comment]
I can revert those visual changes in this PR @nevalla if you want.

**Comment by nevalla** [comment]
> I can revert those visual changes in this PR @nevalla if you want.

Yes, that would be nice, since this is definitely fixing the reported issue.

File: src/renderer/components/+namespaces/namespace-select.tsx (hunk 1 above)
Comment (Nokel81) [reply src/renderer/components/+namespaces/namespace-select.tsx:111]: Yes. For instance "add-quota-dialog".

**Comment by jim-docker** [comment]
> Noticed that this PR also in addition to bug fix, changes the behaviour a bit too. First, there's no more "namespace" icons anymore rendered. Also, when selecting a namespace, the background of selected item is changing. Those changes are a bit out of scope of the actual fix. I'm not saying those changes are bad, but I have a bit mixed feelings changing those in this PR.
> 
> Since this is touching quite many files and fundamentally changes some parts of the system, I would like to have extra eyes too to check this out.

I think the work unrelated to the bug should be moved to another PR. It's not obvious which of these changes actually fix the bug. If we have to come back to this later it will be unnecessarily complicated.

**Comment by Nokel81** [comment]
I'll look through the changes and try and distill them into the necessary fix and then with a subsequent cleanup PR.

**Comment by Nokel81** [comment]
I have reverted `NamespaceSelect`. However I think that changing the namepsace-store to use `Set` is a better model for what the store is supposed to do. Finally the additions of `@action` and `@computed` are necessary for the fix.

**Comment by Nokel81** [comment]
Displaying the diff without the whitespace changes might help when reviewing

**Comment by jim-docker** [comment]
> Displaying the diff without the whitespace changes might help when reviewing

Thanks. Maybe what you can do (next time :) ) is put the improvement refactoring in a single commit, and then have the actual fix in a separate, subsequent commit. That way it is easier to review the fix, plus make separate PRs, if it seems more appropriate.

**Comment by aleksfront** [comment]
@Nokel81 There is a quicker way to fix the "jumping" issue, no need for additional styles and `react select` replacements.
This return statement inside `src/renderer/components/+namespaces/namespace-select.tsx` will to the trick
```
return (
      <NamespaceSelect
        placeholder={placeholder}
        closeMenuOnSelect={false}
        isOptionSelected={() => false}
        controlShouldRenderValue={false}
        isMulti
        onChange={(options: SelectOption[]) => toggleContext(options[0].value)}
        formatOptionLabel={({ value: namespace }: SelectOption) => {
          const isSelected = hasContext(namespace);
          return (
            <div className="flex gaps align-center">
              <FilterIcon type={FilterType.NAMESPACE}/>
              <span>{namespace}</span>
              {isSelected && <Icon small material="check" className="box right"/>}
            </div>
          )
        }}
      />
    )
```
Actually changed 2 lines here - one adding `multyLine` param to `NamespaceSelect`, then parsing "chaged" value according to array retrieved.
namespace selector

My guess also to move any other changes to separate PR or rename/expand current one with additional descriptions.

**Review by aleksfront — CHANGES_REQUESTED** [review]
Requesting changes described in https://github.com/lensapp/lens/pull/911#issuecomment-699888397

**Comment by Nokel81** [comment]
Thanks for the review, I had tried just the `isMulti` but I didn't know about the required `onChange`. Cool.

**Comment by Nokel81** [comment]
Closing this in favour of #987. Will make a follow up PR with some cleaning up of the code later.

**Pull request closed without merging**